### PR TITLE
Addressed WP-737 support ticket: added support for correct log participants tagging in case of user fallback forwarding

### PR DESCRIPTION
**Pull request by DrPyser**

### CODE REVIEW

**Comment by DrPyser** [comment]
> SonarCloud Quality Gate failed.    [Quality Gate failed](https://sonarcloud.io/dashboard?id=wazo-platform_wazo-call-logd&pullRequest=174)
> 
> [Bug](https://sonarcloud.io/project/issues?id=wazo-platform_wazo-call-logd&pullRequest=174&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=wazo-platform_wazo-call-logd&pullRequest=174&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=wazo-platform_wazo-call-logd&pullRequest=174&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=wazo-platform_wazo-call-logd&pullRequest=174&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=wazo-platform_wazo-call-logd&pullRequest=174&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=wazo-platform_wazo-call-logd&pullRequest=174&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=wazo-platform_wazo-call-logd&pullRequest=174&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=wazo-platform_wazo-call-logd&pullRequest=174&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=wazo-platform_wazo-call-logd&pullRequest=174&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=wazo-platform_wazo-call-logd&pullRequest=174&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=wazo-platform_wazo-call-logd&pullRequest=174&resolved=false&types=CODE_SMELL) [27 Code Smells](https://sonarcloud.io/project/issues?id=wazo-platform_wazo-call-logd&pullRequest=174&resolved=false&types=CODE_SMELL)
> 
> [No Coverage information](https://sonarcloud.io/component_measures?id=wazo-platform_wazo-call-logd&pullRequest=174) No Coverage information [](https://sonarcloud.io/component_measures?id=wazo-platform_wazo-call-logd&pullRequest=174&metric=new_duplicated_lines_density&view=list) [13.3% Duplication](https://sonarcloud.io/component_measures?id=wazo-platform_wazo-call-logd&pullRequest=174&metric=new_duplicated_lines_density&view=list)

I don't understand sonarcloud's duplication check. It doesn't seem to explicitly point out which lines are considered duplicated. In this case this is failing on the integration test cases. Test cases can certainly involve much duplication, probably this can be ignored?

**Comment by sboily** [comment]
if you think it's not normal you can ignore it in sonar.

**Comment by DrPyser** [comment]
> @pc-m and I had agreed that updating the API with more information about participants was desirable to consistently/cleanly resolve the use case, but after consideration with @JipiLambert this has been split out into a separate ticket: https://wazo-dev.atlassian.net/browse/WAZO-3067 .

While it would be easy to add a `participants` attribute listing all registered participants(at least their user_uuid), it would be of limited use without additional information on forwarding relationships.
The portal UI would have to be updated to exploit that information as well: https://wazo-dev.atlassian.net/browse/NES-2087 .

File: wazo_call_logd/tests/test_cel_interpretor.py [hunk 1]
```diff
@@ -277,11 +277,14 @@ def _assert_that_interpret_cel_calls(self, function, eventtype):
...
         self.caller_cel_interpretor = CallerCELInterpretor()
-        self.call = Mock(
+        self.call = create_autospec(
             RawCallLog,
+            instance=True,
             interpret_caller_xivo_user_fwd=True,
             extension_filter=Mock(filter=lambda x: x),
+            participants=[],
         )
+        assert hasattr(self.call, "participants")
```
Comment (sduthil) [inline wazo_call_logd/tests/test_cel_interpretor.py:287]: No need for an assert in setup code IMHO

File: integration_tests/suite/test_call_log_generation.py [hunk 2]
```diff
@@ -1721,3 +1722,157 @@ def test_adhoc_conference(self):
...
+1674248467.40 | 1674248467.40 | 2023-01-20 21:01:29.900072+00 | LINKEDID_END              | PJSIP/rgcZLNGE-00000028 | 8000      | 98001      | test-UC-1  | 8000      | unreachable       |                                                                                                                                                                                                                             
+        '''
+    )
+    def test_user_missed_call_fallback_missed(self):
+        # user A missed, fallback to user B, missed too.
+        # Both were called, call log should include both as participants.
+        USER_A_UUID = "ad5b78cf-6e15-45c7-9ef3-bec36e07e8d6"
+        USER_B_UUID = "31be0853-dde6-48cd-986d-85bc708754a1"
+        tenant = "54eb71f8-1f4b-4ae4-8730-638062fbe521"
```
Comment (sduthil) [inline integration_tests/suite/test_call_log_generation.py:1757]: Uppercase or lowercase for test constants, I don't care, but please keep consistent

Comment (sduthil) [reply integration_tests/suite/test_call_log_generation.py:1757]: Also, why not using `USERS_TENANT`? I see that you changed it but I don't understand why.

**Review by sduthil — CHANGES_REQUESTED** [review]
Good job Charles, there are good ideas in there :)

I'll leave in vacation on Friday evening, so I guess somebody else will have to finish the review if we can't finish it tomorrow.

File: integration_tests/suite/test_call_log_generation.py [hunk 3]
```diff
@@ -1,5 +1,6 @@
 # Copyright 2022-2023 The Wazo Authors  (see the AUTHORS file)
 # SPDX-License-Identifier: GPL-3.0-or-later
+# flake8: noqa: W291
```
[the remark this answers is not in the document]
Comment (DrPyser) [reply integration_tests/suite/test_call_log_generation.py:3]: Because flake8 was complaining about trailing spaces in the test data strings. I admit ignoring all trailing whitespaces is maybe not ideal, and I would expect string contents to be excluded.

File: integration_tests/suite/helpers/base.py [hunk 4]
```diff
@@ -131,6 +131,13 @@ def __getattr__(self, member):
 
 
 class _BaseIntegrationTest(AssetLaunchingTestCase):
+    bus: CallLogBusClient
```
[the remark this answers is not in the document]
Comment (DrPyser) [reply integration_tests/suite/helpers/base.py:134]: Yes, of course. Didn't think this was in scope for this PR. I only added minimal annotations for my needs, and I didn't want to get bogged down on debugging all sorts of typing errors/challenges at the same time as addressing the PR's original objective.

File: wazo_call_logd/generator.py [hunk 5]
```diff
@@ -91,57 +89,42 @@ def _remove_duplicate_participants(self, call_log):
             for duplicate_channel_name in duplicate_channel_names:
                 call_log.raw_participants.pop(duplicate_channel_name, None)
 
-    def _fetch_participants(self, confd, call_log):
-        call_log.participants = []
+    def _fetch_participants(self, confd, call_log: raw_call_log.RawCallLog):
+        participants_by_uuid: dict[str, CallLogParticipant] = {
+            str(participant.user_uuid): participant
```
[the remark this answers is not in the document]
Comment (DrPyser) [reply wazo_call_logd/generator.py:94]: Really?
https://github.com/wazo-platform/wazo-call-logd/blob/b88b9fe9e9b01781ad7e1f1bfd4849115190a4e3/wazo_call_logd/database/models.py#L190

The `user_uuid` field is non-nullable. 
I believe all usages of CallLogParticipant imply a `user_uuid`.

[the remark this answers is not in the document]
Comment (DrPyser) [reply wazo_call_logd/generator.py:94]: But I could be safe here and add a check `if participant.user_uuid` in the comprehension.

File: wazo_call_logd/generator.py [hunk 6]
```diff
@@ -91,57 +89,42 @@ def _remove_duplicate_participants(self, call_log):
...
-            )
-
-        if destination_user_uuid:
-            destination_user = confd.users.get(destination_user_uuid)
-            destination_user_tags = (
-                [tag.strip() for tag in destination_user['userfield'].split(',')]
-                if destination_user['userfield']
-                else []
-            )
+            raw_attributes.update(**confd_participant._asdict())
```
[the remark this answers is not in the document]
Comment (DrPyser) [reply wazo_call_logd/generator.py:102]: Yes I know haha.

File: wazo_call_logd/participant.py [hunk 7]
```diff
@@ -55,13 +101,24 @@ def find_participant(confd, channame):
...
-        'tenant_uuid': user['tenant_uuid'],
-        'line_id': line['id'],
-        'tags': tags,
-        'main_extension': main_extension,
-    }
+
+    try:
+        user = confd.users.get(user_uuid)
+    except requests.exceptions.HTTPError:
+        logger.exception("Error retrieving user(user_uuid=%s) from confd", user_uuid)
```
[the remark this answers is not in the document]
Comment (DrPyser) [reply wazo_call_logd/participant.py:108]: Okay. I could add the http status code in the message to make the failure more understandable, instead of the exception trace.

File: wazo_call_logd/raw_call_log.py [hunk 8]
```diff
@@ -17,57 +23,59 @@
...
-        self.recordings = []
-        self.cel_ids = []
-        self.interpret_callee_bridge_enter = True
-        self.interpret_caller_xivo_user_fwd = True
-        self._tenant_uuid = None
-        self.pending_wait_for_mobile_peers = set()
-        self.caller_id_by_channels = {}
-        self.extension_filter = ExtensionFilter(DEFAULT_HIDDEN_EXTENSIONS)
-        self.destination_details = []
+        self.date: datetime = None  # type: ignore[assignment]
```
[the remark this answers is not in the document]
Comment (DrPyser) [reply wazo_call_logd/raw_call_log.py:26]: An Optional field can be None at any point, which means any usage of the field must check for `None` value. Here the field is only initialized to `None` but the value of the field is expected to be set before the field is used.
This is a common case for builder pattern style class. 
I could make all fields Optional, and make sure to check their value and raise exceptions for unset fields in the `to_call_log` method.

File: wazo_call_logd/tests/test_cel_interpretor.py (hunk 1 above)
Comment (DrPyser) [reply wazo_call_logd/tests/test_cel_interpretor.py:287]: I added the assert because the `participants` field wasn't available to the mock(despite being part of the fields in the class interface, which I would expect the mock to know). 
It's not necessary anymore since I set `participants` explicitly in the mock.

File: integration_tests/suite/test_call_log_generation.py (hunk 2 above)
Comment (DrPyser) [reply integration_tests/suite/test_call_log_generation.py:1757]: Because the tenant had to be consistent with the data in the CEL table of the test.

File: wazo_call_logd/generator.py [hunk 9]
```diff
@@ -91,57 +89,42 @@ def _remove_duplicate_participants(self, call_log):
...
-                    participant.line_id = destination_user['lines'][0]['id']
-            call_log_new_participants.append(participant)
-        call_log.participants = call_log_new_participants
+            participant.line_id = confd_participant.line_id
+            participant.tags = confd_participant.tags
+            participant.role = raw_attributes['role']
+            if 'answered' in raw_attributes:
+                participant.answered = raw_attributes["answered"]
+
+        if participants_by_uuid:
```
[the remark this answers is not in the document]
Comment (DrPyser) [reply wazo_call_logd/generator.py:115]: I guess I thought this made it explicit(thus more readable) that the collection might be empty and the following block is conditional on the collection not being empty. But it's true that it is unnecessary.

File: wazo_call_logd/generator.py [hunk 10]
```diff
@@ -91,57 +89,42 @@ def _remove_duplicate_participants(self, call_log):
...
+            # remaining participants identified by CEL interpretation but have no matching channel
+            for uuid, participant in participants_by_uuid.items():
```
[the remark this answers is not in the document]
Comment (DrPyser) [reply wazo_call_logd/generator.py:117]: TL;DR: fetching participants by channel should have priority because of authoritative information on line.

In my mind, there are these cases for participants:
* they have been identified(with their user_uuid) from CEL interpretation
  * they have been reached, so they have an associated channel
  * they have not been reached, so have no associated channel
* they have not been identified from CEL interpretation. they will only be identified from their corresponding channel

However, only if the participant has been reached and has an associated channel can the proper line be associated to the participant. The `find_participant` function first fetches the line from the channel name, then the associated user.
If there is no channel, the `find_participant_by_uuid` fetches the user directly by its `user_uuid`(extracted from CEL interpretation). The implementation provides the first registered line of the user as the value for `line_id`, because that is what was previously done by `_update_call_participants_with_their_tags` for source and destination users. 
This may not be desirable, and `line_id` could be left to `None` for this case.

File: wazo_call_logd/participant.py [hunk 11]
```diff
@@ -1,15 +1,71 @@
...
+
+    tags = get_tags(user['userfield'])
+    logger.debug(
+        'Found participant with user uuid %s, tenant uuid %s',
+        user['uuid'],
+        user['tenant_uuid'],
+    )
+    # NOTE(charles): this replicates the behavior of CallLogsGenerator._update_call_participants_with_their_tags
+    # of defaulting to first line of user, though this might be erroneous
+    line = user['lines'] and user['lines'][0]
```
[the remark this answers is not in the document]
Comment (DrPyser) [reply wazo_call_logd/participant.py:45]: Yes, but is it always guaranteed to be non-null and have at least one value?

File: wazo_call_logd/participant.py [hunk 12]
```diff
@@ -1,15 +1,71 @@
...
+        line_id=line['id'],
+        tags=tags,
+        main_extension=main_extension,
+    )
+
+
+def find_participant(confd: ConfdClient, channame: str) -> ParticipantInfo | None:
+    """
+    find and fetch participant information from confd,
+    using channel name or user_uuid as available
```
[the remark this answers is not in the document]
Comment (DrPyser) [reply wazo_call_logd/participant.py:65]: Oups, sorry. I previously had an alternative interface for this function which accepted a `user_uuid`.

File: wazo_call_logd/participant.py [hunk 13]
```diff
@@ -1,15 +1,71 @@
...
+    # NOTE(charles): this replicates the behavior of CallLogsGenerator._update_call_participants_with_their_tags
+    # of defaulting to first line of user, though this might be erroneous
+    line = user['lines'] and user['lines'][0]
+    logger.debug("user(user_uuid=%s) has first line: %s", user_uuid, line)
+    main_extension = (
+        line
+        and ('extensions' in line or None)
+        and (line['extensions'] or None)
+        and line['extensions'][0]
+    )
```
[the remark this answers is not in the document]
Comment (DrPyser) [reply wazo_call_logd/participant.py:52]: I believe I added the check for the `extensions` key because a test failed on this attribute. I'll see about changing the tests mock instead.

The confd client is untyped and has no schema validation, so at the code level little is guaranteed. Existing tests did not use mock data that matched the API contract. Exporting schema definitions in client libraries would enable more explicit interfaces and guide(perhaps help autogenerate) better mock data for tests.

File: wazo_call_logd/generator.py [hunk 14]
```diff
@@ -91,57 +89,42 @@ def _remove_duplicate_participants(self, call_log):
...
-        if destination_user_uuid:
-            destination_user = confd.users.get(destination_user_uuid)
-            destination_user_tags = (
-                [tag.strip() for tag in destination_user['userfield'].split(',')]
-                if destination_user['userfield']
-                else []
-            )
+            raw_attributes.update(**confd_participant._asdict())
+            if confd_participant.uuid in participants_by_uuid:
+                participant = participants_by_uuid.pop(confd_participant.uuid)
```
[the remark this answers is not in the document]
Comment (DrPyser) [reply wazo_call_logd/generator.py:105]: > As a corner case, a same user can appear twice in the list of participants (though I'm not sure what we want in that case)

The use of mappings prevent iterating twice over the same participant here, no?

File: wazo_call_logd/participant.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (DrPyser) [reply wazo_call_logd/participant.py]: And indeed, integration tests are the reasons for this check of `extensions` key.

File: wazo_call_logd/generator.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (DrPyser) [reply wazo_call_logd/generator.py:105]: It's true that this is an issue I observed, that the same user is "registered" as multiple participants. In this case only one of the participant objects would be updated, which could result in bugs.

[the remark this answers is not in the document]
Comment (DrPyser) [reply wazo_call_logd/generator.py:105]: Oh I misunderstood. A user can logically be multiple participants, such as being the caller *and* the callee, or being present at multiple points of a forwarding chain with cycles or such. This does affect how this is implemented. Will have to work on it.

File: integration_tests/suite/test_call_log_generation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (fblackburn1) [reply integration_tests/suite/test_call_log_generation.py]: @DrPyser I don't understand why removing them is a problem
see: https://github.com/wazo-platform/wazo-call-logd/pull/174/commits/aee29f8829ede36cb1d91eeb0fc454ce85e77514

File: wazo_call_logd/participant.py [hunk 15]
```diff
@@ -1,15 +1,71 @@
...
+
+def find_participant_by_uuid(confd, user_uuid: str) -> ParticipantInfo | None:
+    try:
+        user = confd.users.get(user_uuid)
+    except requests.exceptions.HTTPError:
+        logger.exception("Error retrieving user(user_uuid=%s) from confd", user_uuid)
+        return None
+
+    if not user:
+        return None
```
[the remark this answers is not in the document]
Comment (fblackburn1) [reply wazo_call_logd/participant.py]: It's `GET /users/<uuid>`
If user doesn't exist, it will raise 404

File: wazo_call_logd/participant.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (fblackburn1) [reply wazo_call_logd/participant.py]: Yeah it can be an empty list. So this `user['lines']` LGTM

File: wazo_call_logd/raw_call_log.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (fblackburn1) [reply wazo_call_logd/raw_call_log.py:26]: For info: https://wazo-platform.org/uc-doc/contributors/typing

File: wazo_call_logd/tests/test_generator.py [hunk 16]
```diff
@@ -151,3 +171,240 @@ def _generate_cel_for_call(self, linked_id, cel_count=3):
...
+            [
+                LocalOriginateCELInterpretor(),
+                DispatchCELInterpretor(
+                    CallerCELInterpretor(),
+                    CalleeCELInterpretor(),
+                ),
+            ],
+        )
+
+    def test_unreachable_user(self):
```
Comment (DrPyser) [inline wazo_call_logd/tests/test_generator.py:312]: Having experienced frustration with debugging through integration tests, this is a proof of concept for having functional testing as unit tests to enable a tighter, better test loop and developer experience.
The higher level of complexity and indirection involved in integration tests makes it hard to debug simple errors in the core logic with them.

This example may be more verbose than ideal(though not much more than integration tests), some improvement could be made to that.

**Comment by DrPyser** [comment]
c.f. Sonarcloud duplication errors.
* sucks that that they cannot be processed the same way as other issues(i.e. ignored on a case-by-case basis)
* I believe duplication is a low priority issue as far as code quality goes. Generally removing duplication involves abstraction, which increase complexity, and blindly addressing duplication(e.g. through inheritance or adding meaningless functions or coupled modules) leads to leaky abstractions. 
  A significant threshold of duplication should warrant abstraction, but Sonarcloud does not seem very smart about it.

File: wazo_call_logd/generator.py [hunk 17]
```diff
@@ -91,57 +91,84 @@ def _remove_duplicate_participants(self, call_log):
...
+        # participant information from CEL interpretation can augment participants extracted from channels
+        # and fill in unreachable participants with no corresponding channel
+        unreached_participants = []
+        users_from_cel = (
+            str(participant_info["user_uuid"])
+            for participant_info in call_log.participants_info
+            if "user_uuid" in participant_info
+        )
+        users_from_channels = (str(p.user_uuid) for p in connected_participants)
+        for user_uuid in uniq(chain(users_from_cel, users_from_channels)):
```
Comment (fblackburn1) [inline wazo_call_logd/generator.py:134]: Why not using `set`? Maybe not most efficient, but how much more easier to maintain/read than the custom `uniq` method
```suggestion
        users_from_cel = (
            str(participant_info["user_uuid"])
            for participant_info in call_log.participants_info
            if "user_uuid" in participant_info
        )
        users_from_channels = (str(p.user_uuid) for p in connected_participants)
        all_unique_users = set([*users_from_cel, *users_from_channels])
        for user_uuid in all_unique_users:
```

File: wazo_call_logd/utils.py [hunk 18]
```diff
@@ -0,0 +1,16 @@
+from __future__ import annotations
+from typing import Callable, TypeVar
+from collections.abc import Iterable, Iterator, Hashable
+
+T = TypeVar("T")
+K = TypeVar("K", bound=Hashable)
+
+
+def uniq(iterable: Iterable[T], key: Callable[[T], K] | None = None) -> Iterator[T]:
```
Comment (fblackburn1) [inline wazo_call_logd/utils.py:9]: This kind of library/helper should always have unit tests

File: integration_tests/suite/test_call_log_generation.py [hunk 19]
```diff
@@ -65,7 +65,7 @@ class TestCallLogGeneration(RawCelIntegrationTest):
 CHAN_START   | 2019-08-28 15:29:21.154157 |          |         |         | ycetqvtr | wazo_wait_for_registration | Local/ycetqvtr@wazo_wait_for_registration-00000005;1 |          |                                                       | 1567020561.34 | 1567020560.33 |                                                      |
 APP_START    | 2019-08-28 15:29:21.145952 | Alice    | 1001    | 1001    | s        | user                       | PJSIP/qxqz31sq-00000017                              | Dial     | Local/ycetqvtr@wazo_wait_for_registration,30,         | 1567020560.33 | 1567020560.33 |                                                      |
 CHAN_START   | 2019-08-28 15:29:20.778532 | Alice    | 1001    |         | 1002     | inside                     | PJSIP/qxqz31sq-00000017                              |          |                                                       | 1567020560.33 | 1567020560.33 |                                                      |
-        '''
+        '''  # noqa: W291
```
Comment (fblackburn1) [inline integration_tests/suite/test_call_log_generation.py:68]: Same comment than previous review. Why adding them? Why not just removing empty line at end?

File: integration_tests/suite/test_call_log_generation.py [hunk 20]
```diff
@@ -1721,3 +1728,156 @@ def test_adhoc_conference(self):
                 ),
             ),
         )
+
+    @raw_cels(
+        '''\
+        linkedid      | uniqueid      | eventtime                | eventtype                 | channame                | cid_ani   | cid_dnid   | cid_name   | cid_num   | extra
```
Comment (fblackburn1) [inline integration_tests/suite/test_call_log_generation.py:1734]: Can we remove `cid_any` and `cid_dnid` columns or they are used?

File: integration_tests/suite/test_call_log_generation.py [hunk 21]
```diff
@@ -1721,3 +1728,156 @@ def test_adhoc_conference(self):
...
+                        role='destination',
+                        answered=False,
+                    ),
+                ),
+            ),
+        )
+
+    @raw_cels(
+        '''\
+        linkedid      | uniqueid      | eventtime                | eventtype                 | channame                | cid_ani   | cid_dnid   | cid_name   | cid_num   | extra
```
Comment (fblackburn1) [inline integration_tests/suite/test_call_log_generation.py:1809]: see previous comment about columns

**Review by fblackburn1 — CHANGES_REQUESTED** [review]
[I won't complete my review]
I will let @sduthil complete the review, since he know corner cases/side effect much more than me.
IMHO this solution is a **way** over engineered or maybe the PR contains too much changes completely out of scope of the ticket (ex: black, typing, handling corner cases, tests poc, etc..).
Moreover, the among of necessary comment in the code really worry me about the complexity and the maintainability of this

File: wazo_call_logd/cel_interpretor.py [hunk 22]
```diff
@@ -289,21 +303,52 @@ def interpret_wazo_user_missed_call(self, cel, call):
         ) = _extract_user_missed_call_variables(extra)
 
         if source_user_uuid:
-            source_participant = CallLogParticipant(
-                role='source',
+            call.source_user_uuid = source_user_uuid
+            info = dict(
```
[the remark this answers is not in the document]
Comment (DrPyser) [reply wazo_call_logd/cel_interpretor.py:307]: Stylistic preferences. less quotes.

File: wazo_call_logd/generator.py (hunk 17 above)
Comment (DrPyser) [reply wazo_call_logd/generator.py:134]: I did start out using `set`s, but decided on this solution to maintain ordering of events.

File: integration_tests/suite/test_call_log_generation.py (hunk 20 above)
Comment (DrPyser) [reply integration_tests/suite/test_call_log_generation.py:1734]: I don't think it's a good idea to tailor test data around tests. Tests should be valid if the underlying implementation changes without affecting the functionality and interface. In reality, all data is consumed by the service to generate the logs, however that data is used. Integration tests models the service under test as a black box, the input should be realistic and only the output matters.

**Comment by DrPyser** [comment]
> [I won't complete my review] I will let @sduthil complete the review, since he know corner cases/side effect much more than me. IMHO this solution is a way over engineered or maybe the PR contains too much changes completely out of scope of the ticket (ex: black, typing, handling corner cases, tests poc, etc..). Moreover, the among of necessary comment in the code really worry me about the complexity and the maintainability of this

From my point of view, the reality of call-logd is complex and begets non-trivial solutions. The existing model underlying the implementation is incomplete and does not match reality in a way that enables supporting new use cases well. And this mismatch is error-prone of course. 

From discussion with Seb, the evolution plan for supporting new use cases and improve performance is to insert new CEL events to inject information from the dialplan, and to eventually limit/remove reliance on confd API. Adding new events and correlating those events means adding complexity as well, and as long as not all information is guaranteed to be made available consistently from CEL events(e.g. users identity, line identity, etc) we still need to call confd and merge those two sources of information. 
Moreover the CEL data is dispersed between channel-based events and custom events placed at various points in the dialplan, and correlating caller-side custom events to callee-side channels events or vice versa is non-trivial and error-prone. More events could help give a clearer picture that is easier to correlate, but different categories of situations needing different interpretation remain.

The changes you see is the result of iterating over a non-trivial problem over two sprints while taking into account feedback and ad-hoc actions necessary to address obstacles.
Typing and black are part of the normal development process and can hardly be avoided, and I believe the corner cases explicitly addressed are those relevant to the ticket or relevant to existing tests.

File: integration_tests/suite/test_call_log_generation.py (hunk 19 above)
Comment (DrPyser) [reply integration_tests/suite/test_call_log_generation.py:68]: Could you reproduce this problem while reviewing?

File: wazo_call_logd/cel_interpretor.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (fblackburn1) [reply wazo_call_logd/cel_interpretor.py:307]: We try to keep same style everywhere by using only `{}`.
Moreover is easier for the reader to read `{}` its also perform better

File: integration_tests/suite/test_call_log_generation.py (hunk 20 above)
Comment (fblackburn1) [reply integration_tests/suite/test_call_log_generation.py:1734]: But you already strip other columns from `cel` ...
I understand your point, IMHO it's only a matter of compromise between maintain/readability and theory
```
id          | integer                  |           | not null | nextval('cel_id_seq'::regclass)
 eventtype   | text                     |           | not null | 
 eventtime   | timestamp with time zone |           | not null | 
 userdeftype | text                     |           | not null | 
 cid_name    | text                     |           | not null | 
 cid_num     | text                     |           | not null | 
 cid_ani     | text                     |           | not null | 
 cid_rdnis   | text                     |           | not null | 
 cid_dnid    | text                     |           | not null | 
 exten       | text                     |           | not null | 
 context     | text                     |           | not null | 
 channame    | text                     |           | not null | 
 appname     | text                     |           | not null | 
 appdata     | text                     |           | not null | 
 amaflags    | integer                  |           | not null | 
 accountcode | text                     |           | not null | 
 peeraccount | text                     |           | not null | 
 uniqueid    | text                     |           | not null | 
 linkedid    | text                     |           | not null | 
 userfield   | text                     |           | not null | 
 peer        | text                     |           | not null | 
 extra       | text                     |           |          | 
 call_log_id | integer                  |           |          | 
```

File: wazo_call_logd/cel_interpretor.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (DrPyser) [reply wazo_call_logd/cel_interpretor.py:298]: I'm skeptical of both claims? Why is it easier to read literal syntax with `"` and `:` symbols instead of keyword arguments syntax that is also extremely common in the language?

[the remark this answers is not in the document]
Comment (DrPyser) [reply wazo_call_logd/cel_interpretor.py:298]: The performance argument would matter in a tight loop, but I don't think it justifies a general policy.

File: integration_tests/suite/test_call_log_generation.py (hunk 20 above)
Comment (DrPyser) [reply integration_tests/suite/test_call_log_generation.py]: I agree it's a tradeoff, and since I already agree to go in that direction I'll go "all the way" in this case, but it's a tradeoff that is bound to bite us back. The approach of colocating data with test code has benefits for consistency/understandability of tests, but the obvious cost is readability. If we wanted to optimize for readability perhaps this isn't the right strategy and we would be better off keeping sample data separate.

File: wazo_call_logd/generator.py (hunk 17 above)
Comment (DrPyser) [reply wazo_call_logd/generator.py]: I went back to sets. This solution does not guarantee meaningful ordering anyway, I think we'd have to track order of appearance of events explicitly for that to be consistently possible.

File: integration_tests/suite/test_call_log_generation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (DrPyser) [reply integration_tests/suite/test_call_log_generation.py]: I just considered it worthless busywork. flake8 has no business checking the content of data strings. The point of being able to simply paste sql table is to make it easy to create and maintain tests by minimizing manual data-entering. Why spend time formatting data beyond basic readability?

File: wazo_call_logd/cel_interpretor.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (DrPyser) [reply wazo_call_logd/cel_interpretor.py]: Admitted defeat, this was corrected.

**Comment by sduthil** [comment]
recheck

**Comment by sduthil** [comment]
recheck

File: wazo_call_logd/generator.py [hunk 23]
```diff
@@ -22,6 +23,114 @@
 )
 
 
+class _ParticipantsProcessor:
+    def __init__(self, confd_client: ConfdClient):
+        self.confd: ConfdClient = confd_client
+        self.confd_participants: dict[str, ParticipantInfo] = {}
+
+    def __call__(self, call_log: RawCallLog) -> RawCallLog:
```
Comment (sduthil) [inline wazo_call_logd/generator.py:31]: Not a fan of the `__call__` method. Suggestion: `process(...)`

File: wazo_call_logd/tests/test_generator.py [hunk 24]
```diff
@@ -166,3 +170,63 @@ def _generate_cel_for_call(self, linked_id, cel_count=3):
...
+        }
+        call_log.participants_info = [
+            {
+                "user_uuid": "some-user-uuid",
+                "answered": True
+            }
+        ]
+        self.confd.users.get.side_effect = requests.exceptions.HTTPError()
+        call_log = self.processor(call_log)
+        assert self.confd.mock_calls
```
Comment (sduthil) [inline wazo_call_logd/tests/test_generator.py:200]: No assertion on `call_log`?

File: wazo_call_logd/tests/test_generator.py [hunk 25]
```diff
@@ -166,3 +170,63 @@ def _generate_cel_for_call(self, linked_id, cel_count=3):
...
+        confd_user = mock_data_dict(uuid="some-user-uuid", lines=[mock_data_dict(name="rgcZLNGE", id=1)])
+        confd_line = mock_data_dict(name="rgcZLNGE", id=1, users=[confd_user])
+        self.confd.users.get.return_value = confd_user
+        self.confd.lines.list.return_value = {
+            "items":  [
+                confd_line
+            ]
+        }
+        call_log = self.processor(call_log)
+        self.assertEqual(len(self.confd.users.get.mock_calls), 1) # verify cache effectiveness
```
Comment (sduthil) [inline wazo_call_logd/tests/test_generator.py:225]: You should avoid mixing different assertion methods in the same test (i.e. hamcrest vs unittest).
Also, I don't understand the comment about "cache effectiveness", are we supposed to have multiple calls without the cache, even with one participant only?

File: wazo_call_logd/utils.py [hunk 26]
```diff
@@ -12,3 +14,33 @@ def find(col: Iterable[T], pred: Callable[[T], bool]) -> T | None:
     for x in col:
         if pred(x):
             return x
+
+
+class defaultdict(OrderedDict):
```
Comment (sduthil) [inline wazo_call_logd/utils.py:19]: Suggestion: `OrderedDefaultDict` to avoid confusion with the builtin `defaultdict`

**Review by sduthil — CHANGES_REQUESTED** [review]
Nice work, a few cosmetic changes left and I think we're good to go :)

File: wazo_call_logd/tests/test_generator.py (hunk 25 above)
Comment (DrPyser) [reply wazo_call_logd/tests/test_generator.py:225]: Yes, because there is a `raw_participant` with a line corresponding to the same user in `participants_info`. Without a cache, `find_participant` would be called to find the user through the line, and then `find_participant_by_uuid` called on the `user_uuid` from `participants_info`.

File: wazo_call_logd/generator.py (hunk 23 above)
Comment (DrPyser) [reply wazo_call_logd/generator.py:31]: In my mind, if a class exists only to provide encapsulation of a single method, and so is equivalent to(could be replaced by) a function closure(some environment + some parameterized code using that environment), it deserves a callable interface.

**Review by sduthil — APPROVED** [review]
